scipy/linalg/basic.py

```diff
@@ -1233,83 +1233,121 @@ def lstsq(a, b, cond=None, overwrite_a=False, overwrite_b=False,
         (crashes, non-termination) if the inputs do contain infinities or NaNs.
+    cond, rcond : float, optional
+        In older versions, these values were meant to be used as ``atol`` with
+        ``rtol=0``. If both were given ``rcond`` overwrote ``cond`` and hence
+        the code was not correct. Thus using these are strongly discouraged and
+        the tolerances above are recommended instead. In fact, if provided,
+        atol, rtol takes precedence over these keywords.
+
+        .. versionchanged:: 1.7.0
+            Deprecated in favor of ``rtol`` and ``atol`` parameters above and
+            will be removed in future versions of SciPy.
+
+        .. versionchanged:: 1.3.0
+            Previously the default cutoff value was just ``eps*f`` where ``f``
+            was ``1e3`` for single precision and ``1e6`` for double precision.
 
     Returns
     -------
     B : (N, M) ndarray
         The pseudo-inverse of matrix `a`.
     rank : int
-        The effective rank of the matrix. Returned if return_rank == True
+        The effective rank of the matrix. Returned if `return_rank` is True.
 
     Raises
     ------
     LinAlgError
-        If computation does not converge.
+        If SVD computation does not converge.
 
     Examples
     --------
     >>> from scipy import linalg
     >>> rng = np.random.default_rng()
     >>> a = rng.standard_normal((9, 6))
     >>> B = linalg.pinv(a)
-    >>> np.allclose(a, np.dot(a, np.dot(B, a)))
+    >>> np.allclose(a, a @ B @ a)
     True
-    >>> np.allclose(B, np.dot(B, np.dot(a, B)))
+    >>> np.allclose(B, B @ a @ B)
     True
 
     """
     a = _asarray_validated(a, check_finite=check_finite)
-    # If a is sufficiently tall it is cheaper to compute using the transpose
-    trans = a.shape[0] / a.shape[1] >= 1.1
-    b = np.eye(a.shape[1] if trans else a.shape[0], dtype=a.dtype)
+    u, s, vh = decomp_svd.svd(a, full_matrices=False, check_finite=False)
+    t = u.dtype.char.lower()
+    maxS = np.max(s)
 
-    if rcond is not None:
-        cond = rcond
+    if rcond or cond:
+        warn('Use of the "cond" and "rcond" keywords are deprecated and '
+             'will be removed in future versions of SciPy. Use "atol" and '
+             '"rtol" keywords instead', DeprecationWarning, stacklevel=2)
 
-    if cond is None:
-        cond = max(a.shape) * np.spacing(a.real.dtype.type(1))
+    # backwards compatible only atol and rtol are both missing
+    if (rcond or cond) and (atol is None) and (rtol is None):
+        atol = rcond or cond
+        rtol = 0.
+
+    atol = 0. if atol is None else atol
+    rtol = max(a.shape) * np.finfo(t).eps if (rtol is None) else rtol
 
-    x, resids, rank, s = lstsq(a.T if trans else a, b,
-                               cond=cond, check_finite=False)
+    if (atol < 0.) or (rtol < 0.):
+        raise ValueError("atol and rtol values must be positive.")
+
+    val = atol + maxS * rtol
```

Comment:
This would help scikit-learn recreate the behavior in SciPy < 1.3.0 by:

```python
t = a.dtype.char.lower()
factor = {'f': 1E3, 'd': 1E6}
rtol = factor[t] * np.finfo(t).eps

pinv(a, rtol=rtol)

Reply:
Yes now it will give everyone some finer control I hope. Do you want that behavior out of the box or can you accomodate it externally?

Reply:
I can accommodate it externally.